File: code/python/04-rag-pipeline/embedding_sync.py
"""Embedding sync manager: keep a vector index in sync with source documents.

The vector database is a *derived index*, not the source of truth.  Documents
change, get deleted, and new ones are added.  When the source document store
drifts from the vector index, search quality degrades silently.

This module provides:

* :class:`EmbeddingSyncManager` — orchestrates full and incremental syncs.
* :class:`InMemoryDocumentStore` — a lightweight document store for demos and
  tests; swap it for your real document storage.
* :class:`SyncReport` / :class:`SyncHealth` — structured results.

Usage::

    from embedding_sync import EmbeddingSyncManager, InMemoryDocumentStore
    from vector_database import SimpleVectorStore

    store = InMemoryDocumentStore()
    db = SimpleVectorStore()
    embedder = lambda text: [0.1] * 64  # replace with real embedder

    manager = EmbeddingSyncManager(db, embedder, store)
    report = manager.full_sync()
    print(report)

    health = manager.verify_sync()
    print(health)

See: docs/05-the-tool-ecosystem/02-vector-databases.md
"""

from __future__ import annotations

import hashlib
import random
import threading
import time
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Callable, Optional

from vector_database import VectorDatabase, VectorDocument


# ---------------------------------------------------------------------------
# Data classes
# ---------------------------------------------------------------------------


@dataclass
class SourceDocument:
    """A document in the source document store."""

    id: str
    text: str
    metadata: dict = field(default_factory=dict)
    updated_at: float = field(default_factory=time.time)

    def content_hash(self) -> str:
        """SHA-256 of the document text — used to detect changes."""
        return hashlib.sha256(self.text.encode()).hexdigest()


@dataclass
class SyncReport:
    """Result of a full or incremental sync operation."""

    strategy: str          # "full" or "incremental"
    created: int = 0
    updated: int = 0
    deleted: int = 0
    errors: list[str] = field(default_factory=list)
    duration_seconds: float = 0.0
    timestamp: datetime = field(default_factory=lambda: datetime.now(timezone.utc))

    def __str__(self) -> str:
        status = "OK" if not self.errors else f"{len(self.errors)} errors"
        return (
            f"SyncReport[{self.strategy}] "
            f"created={self.created} updated={self.updated} deleted={self.deleted} "
            f"duration={self.duration_seconds:.2f}s status={status} "
            f"at={self.timestamp.isoformat()}"
        )


@dataclass
class SyncHealth:
    """Result of a sync health verification."""

    is_healthy: bool
    total_documents: int      # Documents in the source store
    total_vectors: int        # Vectors in the vector DB
    mismatch_count: int       # Documents in store but not in vector DB
    orphan_count: int         # Vectors in DB with no corresponding document
    recommendations: list[str] = field(default_factory=list)

    def __str__(self) -> str:
        status = "HEALTHY" if self.is_healthy else "DEGRADED"
        return (
            f"SyncHealth[{status}] "
            f"docs={self.total_documents} vectors={self.total_vectors} "
            f"missing={self.mismatch_count} orphans={self.orphan_count} "
            f"recommendations={self.recommendations}"
        )


# ---------------------------------------------------------------------------
# Simple in-memory document store (for demos and tests)
# ---------------------------------------------------------------------------


class InMemoryDocumentStore:
    """Minimal in-memory document store.

    Tracks creation/update timestamps so incremental sync can query for
    documents changed after a given timestamp.
    """

    def __init__(self) -> None:
        self._docs: dict[str, SourceDocument] = {}

    # -- Mutation --

    def add(self, doc: SourceDocument) -> None:
        """Add or replace a document."""
        self._docs[doc.id] = doc

    def update(self, doc_id: str, new_text: str, metadata: Optional[dict] = None) -> None:
        """Update the text (and optionally metadata) of an existing document."""
        existing = self._docs.get(doc_id)
        if existing is None:
            raise KeyError(f"Document {doc_id!r} not found")
        self._docs[doc_id] = SourceDocument(
            id=doc_id,
            text=new_text,
            metadata=metadata if metadata is not None else existing.metadata,
            updated_at=time.time(),
        )

    def remove(self, doc_id: str) -> None:
        """Remove a document by ID."""
        self._docs.pop(doc_id, None)

    def add_many(self, docs: list[SourceDocument]) -> None:
        for doc in docs:
            self.add(doc)

    # -- Query --

    def get_all(self) -> list[SourceDocument]:
        return list(self._docs.values())

    def get_ids(self) -> set[str]:
        return set(self._docs)

    def get_changes(
        self, since: Optional[float] = None
    ) -> dict[str, list]:
        """Return documents created or updated after *since* (epoch seconds).

        Returns a dict with keys ``"created_or_updated"`` and ``"all_ids"``
        (all current document IDs, used to find deleted vectors).
        """
        if since is None:
            return {"created_or_updated": self.get_all(), "all_ids": self.get_ids()}
        changed = [d for d in self._docs.values() if d.updated_at > since]
        return {"created_or_updated": changed, "all_ids": self.get_ids()}

    def count(self) -> int:
        return len(self._docs)


# ---------------------------------------------------------------------------
# EmbeddingSyncManager
# ---------------------------------------------------------------------------

# Embedder type: any callable that takes a string and returns a list of floats
Embedder = Callable[[str], list[float]]


class EmbeddingSyncManager:
    """Synchronise a vector database index with a source document store.

    The manager supports two sync strategies:

    * **Full sync** — clear the vector index and re-embed every document from
      scratch.  Use this for initial population, after changing the embedding
      model, or after a major schema change.
    * **Incremental sync** — only process documents added, updated, or deleted
      since the last sync timestamp.  Suitable for regular background jobs.

    A built-in scheduler can run either strategy on a recurring timer.

    Args:
        vector_db:      The vector database to keep in sync.
        embedder:       Callable that converts text to an embedding vector.
        document_store: The authoritative source of documents.
    """

    def __init__(
        self,
        vector_db: VectorDatabase,
        embedder: Embedder,
        document_store: InMemoryDocumentStore,
    ) -> None:
        self.vector_db = vector_db
        self.embedder = embedder
        self.documents = document_store

        self._last_sync_ts: Optional[float] = None
        self._scheduler_thread: Optional[threading.Thread] = None
        self._scheduler_stop = threading.Event()

    # ------------------------------------------------------------------
    # Full sync
    # ------------------------------------------------------------------

    def full_sync(self) -> SyncReport:
        """Re-embed all documents and rebuild the entire vector index.

        Existing vectors are deleted before insertion.

        Returns:
            A :class:`SyncReport` describing what happened.
        """
        t0 = time.perf_counter()
        errors: list[str] = []
        created = 0

        self.vector_db.clear()

        for doc in self.documents.get_all():
            try:
                vdoc = self._to_vector_document(doc)
                self.vector_db.insert([vdoc])
                created += 1
            except Exception as exc:
                errors.append(f"[{doc.id}] {exc}")

        self._last_sync_ts = time.time()
        return SyncReport(
            strategy="full",
            created=created,
            updated=0,
            deleted=0,
            errors=errors,
            duration_seconds=time.perf_counter() - t0,
        )

    # ------------------------------------------------------------------
    # Incremental sync
    # ------------------------------------------------------------------

    def incremental_sync(self, since: Optional[float] = None) -> SyncReport:
        """Only sync documents that have changed since *since*.

        Args:
            since: Epoch timestamp.  Defaults to the timestamp of the last
                   sync (or the beginning of time if no sync has run yet).

        Returns:
            A :class:`SyncReport` describing what happened.
        """
        since = since if since is not None else self._last_sync_ts
        t0 = time.perf_counter()
        errors: list[str] = []
        created = updated = deleted = 0

        changes = self.documents.get_changes(since=since)
        current_ids = changes["all_ids"]

        # --- Handle deletions: vectors with no matching source document ---
        # We detect deletions by querying the vector DB for IDs that no longer
        # appear in the document store.  SimpleVectorStore exposes _docs;
        # for production backends you would maintain a separate ID set.
        vector_ids = self._get_vector_ids()
        deleted_ids = vector_ids - current_ids
        if deleted_ids:
            try:
                self.vector_db.delete(list(deleted_ids))
                deleted = len(deleted_ids)
            except Exception as exc:
                errors.append(f"delete batch: {exc}")

        # --- Handle additions and updates ---
        for doc in changes["created_or_updated"]:
            try:
                vdoc = self._to_vector_document(doc)
                # Upsert: delete old embedding then insert fresh one
                if doc.id in vector_ids:
                    self.vector_db.delete([doc.id])
                    self.vector_db.insert([vdoc])
                    updated += 1
                else:
                    self.vector_db.insert([vdoc])
                    created += 1
            except Exception as exc:
                errors.append(f"[{doc.id}] {exc}")

        self._last_sync_ts = time.time()
        return SyncReport(
            strategy="incremental",
            created=created,
            updated=updated,
            deleted=deleted,
            errors=errors,
            duration_seconds=time.perf_counter() - t0,
        )

    # ------------------------------------------------------------------
    # Verify
    # ------------------------------------------------------------------

    def verify_sync(self, sample_size: int = 100) -> SyncHealth:
        """Check whether the vector index is consistent with the document store.

        Randomly samples up to *sample_size* documents, checks that each has
        a corresponding vector, and checks for orphaned vectors.

        Args:
            sample_size: Maximum number of documents to sample.

        Returns:
            A :class:`SyncHealth` report.
        """
        all_docs = self.documents.get_all()
        total_docs = len(all_docs)
        total_vectors = self.vector_db.count()

        sample = random.sample(all_docs, min(sample_size, total_docs))
        vector_ids = self._get_vector_ids()

        mismatch_count = 0
        for doc in sample:
            if doc.id not in vector_ids:
                mismatch_count += 1

        # Orphans: in vector DB but not in store
        store_ids = self.documents.get_ids()
        orphan_count = len(vector_ids - store_ids)

        recommendations: list[str] = []
        if mismatch_count > 0:
            recommendations.append(
                f"{mismatch_count} sampled documents missing from vector DB — "
                "run incremental_sync() or full_sync()."
            )
        if orphan_count > 0:
            recommendations.append(
                f"{orphan_count} orphaned vectors found — "
                "run incremental_sync() to clean up."
            )
        if total_docs != total_vectors:
            recommendations.append(
                f"Document count ({total_docs}) ≠ vector count ({total_vectors}) — "
                "index may be stale."
            )

        is_healthy = mismatch_count == 0 and orphan_count == 0 and total_docs == total_vectors
        return SyncHealth(
            is_healthy=is_healthy,
            total_documents=total_docs,
            total_vectors=total_vectors,
            mismatch_count=mismatch_count,
            orphan_count=orphan_count,
            recommendations=recommendations,
        )

    # ------------------------------------------------------------------
    # Scheduler
    # ------------------------------------------------------------------

    def schedule_sync(
        self,
        strategy: str = "incremental",
        interval_minutes: int = 15,
    ) -> None:
        """Start a background thread that runs syncs on a timer.

        Strategies:

        * ``"incremental"`` — runs every *interval_minutes*.
        * ``"full"`` — runs every *interval_minutes*.
        * ``"verify"`` — runs :meth:`verify_sync` every *interval_minutes*
          and prints the health report (does not modify the index).

        Call :meth:`stop_scheduler` to cancel the background thread.

        Args:
            strategy:         ``"incremental"``, ``"full"``, or ``"verify"``.
            interval_minutes: Seconds between runs (reuses the name for clarity).
        """
        if self._scheduler_thread and self._scheduler_thread.is_alive():
            raise RuntimeError("A scheduler is already running.  Call stop_scheduler() first.")

        self._scheduler_stop.clear()

        def _loop() -> None:
            while not self._scheduler_stop.wait(timeout=interval_minutes * 60):
                if strategy == "full":
                    report = self.full_sync()
                    print(f"[Scheduler] {report}")
                elif strategy == "verify":
                    health = self.verify_sync()
                    print(f"[Scheduler] {health}")
                else:
                    report = self.incremental_sync()
                    print(f"[Scheduler] {report}")

        self._scheduler_thread = threading.Thread(target=_loop, daemon=True)
        self._scheduler_thread.start()

    def stop_scheduler(self) -> None:
        """Stop the background sync scheduler."""
        self._scheduler_stop.set()
        if self._scheduler_thread:
            self._scheduler_thread.join(timeout=5)

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _to_vector_document(self, doc: SourceDocument) -> VectorDocument:
        embedding = self.embedder(doc.text)
        return VectorDocument(
            id=doc.id,
            text=doc.text,
            embedding=embedding,
            metadata={
                **doc.metadata,
                "_content_hash": doc.content_hash(),
                "_updated_at": doc.updated_at,
            },
        )

    def _get_vector_ids(self) -> set[str]:
        """Return the set of IDs currently stored in the vector database.

        For SimpleVectorStore we access ``_docs`` directly.  For production
        backends you would maintain a side-channel ID set or use the
        database's listing API.
        """
        # SimpleVectorStore exposes _docs
        if hasattr(self.vector_db, "_docs"):
            return {d["id"] for d in self.vector_db._docs}
        # Generic fallback: we can't cheaply enumerate all IDs
        # Production implementations should override this or use a metadata store.
        return set()


# ---------------------------------------------------------------------------
# Demo
# ---------------------------------------------------------------------------


def _run_demo() -> None:
    """Create 50 documents, run full sync, apply changes, then incremental sync
    and verify health.  Also demonstrates a simulated embedding model change."""
    import numpy as np
    from vector_database import SimpleVectorStore

    rng = np.random.default_rng(7)
    DIM = 64

    def make_embedder(seed: int) -> Embedder:
        """Return a deterministic embedder seeded to *seed*."""
        r = np.random.default_rng(seed)

        def _embed(text: str) -> list[float]:
            # Deterministic: hash text to a seed offset, then draw from rng
            offset = int(hashlib.md5(text.encode()).hexdigest(), 16) % 1_000_000
            local = np.random.default_rng(seed + offset)
            v = local.standard_normal(DIM)
            return (v / np.linalg.norm(v)).tolist()

        return _embed

    embedder_v1 = make_embedder(42)

    # --- Build document store ---
    store = InMemoryDocumentStore()
    for i in range(50):
        store.add(
            SourceDocument(
                id=f"doc_{i:03d}",
                text=f"Document {i}: information about topic {i % 10}.",
                metadata={"category": f"cat_{i % 5}"},
            )
        )

    db = SimpleVectorStore()
    manager = EmbeddingSyncManager(db, embedder_v1, store)

    print("=" * 60)
    print("Embedding Sync Demo")
    print("=" * 60)

    # --- Full sync ---
    report = manager.full_sync()
    print(f"\n[1] Full sync:        {report}")
    health = manager.verify_sync(sample_size=50)
    print(f"    Health check:     {health}")

    # --- Modify documents ---
    time.sleep(0.01)  # ensure updated_at is strictly later
    for i in range(5):
        store.update(f"doc_{i:03d}", f"UPDATED — Document {i} revised content.")
    for i in range(45, 48):
        store.remove(f"doc_{i:03d}")
    for i in range(50, 52):
        store.add(
            SourceDocument(
                id=f"doc_{i:03d}",
                text=f"New document {i} added after initial sync.",
            )
        )

    # --- Incremental sync ---
    report2 = manager.incremental_sync()
    print(f"\n[2] Incremental sync: {report2}")
    health2 = manager.verify_sync(sample_size=50)
    print(f"    Health check:     {health2}")

    # --- Simulate embedding model change ---
    print("\n[3] Simulating embedding model change (v1 → v2) ...")
    embedder_v2 = make_embedder(99)  # different seed = different vector space
    manager.embedder = embedder_v2
    report3 = manager.full_sync()
    print(f"    Full re-sync:     {report3}")
    health3 = manager.verify_sync(sample_size=50)
    print(f"    Health check:     {health3}")


if __name__ == "__main__":
    _run_demo()

File: code/go/04-rag-pipeline/embedding_sync.go
// embedding_sync.go — Keep a vector index in sync with source documents.
//
// Provides full and incremental sync strategies plus a health verification check.
// Uses the VectorDatabase interface so it works with any backend.
//
// See: docs/05-the-tool-ecosystem/02-vector-databases.md
package ragpipeline

import (
	"crypto/sha256"
	"fmt"
	"math/rand"
	"sync"
	"time"
)

// ---------------------------------------------------------------------------
// SourceDocument
// ---------------------------------------------------------------------------

// SourceDocument is a document in the source document store.
type SourceDocument struct {
	ID        string
	Text      string
	Metadata  map[string]interface{}
	UpdatedAt float64 // Unix timestamp
}

// ContentHash returns the SHA-256 hex digest of the document text.
func (d *SourceDocument) ContentHash() string {
	sum := sha256.Sum256([]byte(d.Text))
	return fmt.Sprintf("%x", sum)
}

// ---------------------------------------------------------------------------
// SyncReport
// ---------------------------------------------------------------------------

// SyncReport is the result of a full or incremental sync operation.
type SyncReport struct {
	Strategy        string
	Created         int
	Updated         int
	Deleted         int
	Errors          []string
	DurationSeconds float64
	Timestamp       time.Time
}

func (r *SyncReport) String() string {
	status := "OK"
	if len(r.Errors) > 0 {
		status = fmt.Sprintf("%d errors", len(r.Errors))
	}
	return fmt.Sprintf("SyncReport[%s] created=%d updated=%d deleted=%d duration=%.2fs status=%s at=%s",
		r.Strategy, r.Created, r.Updated, r.Deleted,
		r.DurationSeconds, status, r.Timestamp.Format(time.RFC3339))
}

// ---------------------------------------------------------------------------
// SyncHealth
// ---------------------------------------------------------------------------

// SyncHealth is the result of a sync health verification.
type SyncHealth struct {
	IsHealthy       bool
	TotalDocuments  int
	TotalVectors    int
	MismatchCount   int
	OrphanCount     int
	Recommendations []string
}

func (h *SyncHealth) String() string {
	status := "HEALTHY"
	if !h.IsHealthy {
		status = "DEGRADED"
	}
	return fmt.Sprintf("SyncHealth[%s] docs=%d vectors=%d missing=%d orphans=%d recommendations=%v",
		status, h.TotalDocuments, h.TotalVectors, h.MismatchCount, h.OrphanCount, h.Recommendations)
}

// ---------------------------------------------------------------------------
// InMemoryDocumentStore
// ---------------------------------------------------------------------------

// InMemoryDocumentStore is a lightweight document store for demos and tests.
type InMemoryDocumentStore struct {
	mu   sync.RWMutex
	docs map[string]SourceDocument
}

// NewInMemoryDocumentStore creates an empty InMemoryDocumentStore.
func NewInMemoryDocumentStore() *InMemoryDocumentStore {
	return &InMemoryDocumentStore{docs: make(map[string]SourceDocument)}
}

// Add adds or replaces a document.
func (s *InMemoryDocumentStore) Add(doc SourceDocument) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if doc.UpdatedAt == 0 {
		doc.UpdatedAt = float64(time.Now().UnixNano()) / 1e9
	}
	s.docs[doc.ID] = doc
}

// AddMany adds multiple documents.
func (s *InMemoryDocumentStore) AddMany(docs []SourceDocument) {
	for _, d := range docs {
		s.Add(d)
	}
}

// Update updates the text (and optionally metadata) of an existing document.
func (s *InMemoryDocumentStore) Update(docID, newText string, metadata map[string]interface{}) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	existing, ok := s.docs[docID]
	if !ok {
		return fmt.Errorf("document %q not found", docID)
	}
	if metadata == nil {
		metadata = existing.Metadata
	}
	s.docs[docID] = SourceDocument{
		ID:        docID,
		Text:      newText,
		Metadata:  metadata,
		UpdatedAt: float64(time.Now().UnixNano()) / 1e9,
	}
	return nil
}

// Remove removes a document by ID.
func (s *InMemoryDocumentStore) Remove(docID string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	delete(s.docs, docID)
}

// GetAll returns all documents.
func (s *InMemoryDocumentStore) GetAll() []SourceDocument {
	s.mu.RLock()
	defer s.mu.RUnlock()
	docs := make([]SourceDocument, 0, len(s.docs))
	for _, d := range s.docs {
		docs = append(docs, d)
	}
	return docs
}

// GetIDs returns the set of all current document IDs.
func (s *InMemoryDocumentStore) GetIDs() map[string]bool {
	s.mu.RLock()
	defer s.mu.RUnlock()
	ids := make(map[string]bool, len(s.docs))
	for id := range s.docs {
		ids[id] = true
	}
	return ids
}

// GetChanges returns documents created or updated after since (epoch seconds).
// since=0 returns all documents.
type StoreChanges struct {
	CreatedOrUpdated []SourceDocument
	AllIDs           map[string]bool
}

func (s *InMemoryDocumentStore) GetChanges(since float64) StoreChanges {
	s.mu.RLock()
	defer s.mu.RUnlock()
	allIDs := make(map[string]bool, len(s.docs))
	var changed []SourceDocument
	for id, doc := range s.docs {
		allIDs[id] = true
		if since == 0 || doc.UpdatedAt > since {
			changed = append(changed, doc)
		}
	}
	return StoreChanges{CreatedOrUpdated: changed, AllIDs: allIDs}
}

// Count returns the number of documents.
func (s *InMemoryDocumentStore) Count() int {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return len(s.docs)
}

// ---------------------------------------------------------------------------
// EmbeddingSyncManager
// ---------------------------------------------------------------------------

// DocEmbedder is any function that converts text to an embedding vector.
type DocEmbedder func(text string) ([]float64, error)

// EmbeddingSyncManager synchronises a VectorDatabase with a document store.
type EmbeddingSyncManager struct {
	VectorDB   VectorDatabase
	Embedder   DocEmbedder
	Documents  *InMemoryDocumentStore
	lastSyncTS float64

	stopCh   chan struct{}
	stopOnce sync.Once
}

// NewEmbeddingSyncManager creates an EmbeddingSyncManager.
func NewEmbeddingSyncManager(db VectorDatabase, embedder DocEmbedder, store *InMemoryDocumentStore) *EmbeddingSyncManager {
	return &EmbeddingSyncManager{
		VectorDB:  db,
		Embedder:  embedder,
		Documents: store,
	}
}

// FullSync re-embeds all documents and rebuilds the entire vector index.
func (m *EmbeddingSyncManager) FullSync() (*SyncReport, error) {
	t0 := time.Now()
	var errors []string
	created := 0

	if err := m.VectorDB.Clear(); err != nil {
		return nil, fmt.Errorf("clear: %w", err)
	}

	for _, doc := range m.Documents.GetAll() {
		vdoc, err := m.toVecDocument(doc)
		if err != nil {
			errors = append(errors, fmt.Sprintf("[%s] %v", doc.ID, err))
			continue
		}
		if _, err := m.VectorDB.Insert([]VecDocument{vdoc}); err != nil {
			errors = append(errors, fmt.Sprintf("[%s] %v", doc.ID, err))
			continue
		}
		created++
	}

	m.lastSyncTS = float64(time.Now().UnixNano()) / 1e9
	return &SyncReport{
		Strategy:        "full",
		Created:         created,
		Errors:          errors,
		DurationSeconds: time.Since(t0).Seconds(),
		Timestamp:       time.Now().UTC(),
	}, nil
}

// IncrementalSync only syncs documents that have changed since the last sync.
func (m *EmbeddingSyncManager) IncrementalSync(since float64) (*SyncReport, error) {
	if since == 0 {
		since = m.lastSyncTS
	}
	t0 := time.Now()
	var errors []string
	created, updated, deleted := 0, 0, 0

	changes := m.Documents.GetChanges(since)
	currentIDs := changes.AllIDs

	// Detect deletions.
	vectorIDs := m.getVectorIDs()
	for id := range vectorIDs {
		if !currentIDs[id] {
			if _, err := m.VectorDB.Delete([]string{id}); err != nil {
				errors = append(errors, fmt.Sprintf("delete %s: %v", id, err))
			} else {
				deleted++
			}
		}
	}

	// Handle additions and updates.
	for _, doc := range changes.CreatedOrUpdated {
		vdoc, err := m.toVecDocument(doc)
		if err != nil {
			errors = append(errors, fmt.Sprintf("[%s] %v", doc.ID, err))
			continue
		}
		if vectorIDs[doc.ID] {
			m.VectorDB.Delete([]string{doc.ID})
			m.VectorDB.Insert([]VecDocument{vdoc})
			updated++
		} else {
			m.VectorDB.Insert([]VecDocument{vdoc})
			created++
		}
	}

	m.lastSyncTS = float64(time.Now().UnixNano()) / 1e9
	return &SyncReport{
		Strategy:        "incremental",
		Created:         created,
		Updated:         updated,
		Deleted:         deleted,
		Errors:          errors,
		DurationSeconds: time.Since(t0).Seconds(),
		Timestamp:       time.Now().UTC(),
	}, nil
}

// VerifySync checks whether the vector index is consistent with the document store.
func (m *EmbeddingSyncManager) VerifySync(sampleSize int) (*SyncHealth, error) {
	allDocs := m.Documents.GetAll()
	totalDocs := len(allDocs)
	totalVectors, err := m.VectorDB.Count()
	if err != nil {
		return nil, err
	}

	if sampleSize > totalDocs {
		sampleSize = totalDocs
	}
	sample := make([]SourceDocument, len(allDocs))
	copy(sample, allDocs)
	rand.Shuffle(len(sample), func(i, j int) { sample[i], sample[j] = sample[j], sample[i] })
	sample = sample[:sampleSize]

	vectorIDs := m.getVectorIDs()
	mismatchCount := 0
	for _, doc := range sample {
		if !vectorIDs[doc.ID] {
			mismatchCount++
		}
	}

	storeIDs := m.Documents.GetIDs()
	orphanCount := 0
	for id := range vectorIDs {
		if !storeIDs[id] {
			orphanCount++
		}
	}

	var recommendations []string
	if mismatchCount > 0 {
		recommendations = append(recommendations, fmt.Sprintf(
			"%d sampled documents missing from vector DB — run IncrementalSync() or FullSync().", mismatchCount))
	}
	if orphanCount > 0 {
		recommendations = append(recommendations, fmt.Sprintf(
			"%d orphaned vectors found — run IncrementalSync() to clean up.", orphanCount))
	}
	if totalDocs != totalVectors {
		recommendations = append(recommendations, fmt.Sprintf(
			"Document count (%d) ≠ vector count (%d) — index may be stale.", totalDocs, totalVectors))
	}

	return &SyncHealth{
		IsHealthy:       mismatchCount == 0 && orphanCount == 0 && totalDocs == totalVectors,
		TotalDocuments:  totalDocs,
		TotalVectors:    totalVectors,
		MismatchCount:   mismatchCount,
		OrphanCount:     orphanCount,
		Recommendations: recommendations,
	}, nil
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

func (m *EmbeddingSyncManager) toVecDocument(doc SourceDocument) (VecDocument, error) {
	emb, err := m.Embedder(doc.Text)
	if err != nil {
		return VecDocument{}, err
	}
	meta := make(map[string]interface{}, len(doc.Metadata)+2)
	for k, v := range doc.Metadata {
		meta[k] = v
	}
	meta["_content_hash"] = doc.ContentHash()
	meta["_updated_at"] = doc.UpdatedAt
	return VecDocument{
		ID:        doc.ID,
		Text:      doc.Text,
		Embedding: emb,
		Metadata:  meta,
	}, nil
}

func (m *EmbeddingSyncManager) getVectorIDs() map[string]bool {
	// InMemoryVecStore (implements VectorDatabase) stores docs internally;
	// we access it via a type assertion.
	if store, ok := m.VectorDB.(*InMemoryVecStore); ok {
		store.mu.RLock()
		defer store.mu.RUnlock()
		ids := make(map[string]bool, len(store.docs))
		for _, d := range store.docs {
			ids[d.id] = true
		}
		return ids
	}
	return map[string]bool{}
}

// ---------------------------------------------------------------------------
// Demo
// ---------------------------------------------------------------------------

// RunEmbeddingSync demonstrates full and incremental sync operations.
func RunEmbeddingSync() {
	fmt.Println("EMBEDDING SYNC DEMO")
	fmt.Println("Set OPENAI_API_KEY to run a live embedding demo.")
	fmt.Println("Using deterministic fake embedder for illustration.")

	fakeEmbedder := func(text string) ([]float64, error) {
		v := make([]float64, 8)
		for i, c := range text {
			v[i%8] += float64(c)
		}
		// Normalize
		norm := 0.0
		for _, x := range v {
			norm += x * x
		}
		if norm > 0 {
			norm = 1.0 / (norm * norm)
			for i := range v {
				v[i] *= norm
			}
		}
		return v, nil
	}

	store := NewInMemoryDocumentStore()
	for i := 0; i < 10; i++ {
		store.Add(SourceDocument{
			ID:   fmt.Sprintf("doc_%03d", i),
			Text: fmt.Sprintf("Document %d: information about topic %d.", i, i%3),
		})
	}

	db := NewInMemoryVecStore()
	manager := NewEmbeddingSyncManager(db, fakeEmbedder, store)

	report, _ := manager.FullSync()
	fmt.Printf("Full sync: %s\n", report)

	store.Add(SourceDocument{ID: "doc_new", Text: "A brand new document."})
	store.Remove("doc_000")

	report2, _ := manager.IncrementalSync(0)
	fmt.Printf("Incremental sync: %s\n", report2)

	health, _ := manager.VerifySync(100)
	fmt.Printf("Health: %s\n", health)
}
